co2calculator/calculate.py

```diff
@@ -21,78 +24,149 @@
-                        e.g.,  [{"address": "Im Neuenheimer Feld 348",
-                                "locality": "Heidelberg",
-                                 "country": "Germany"},
-                                 {"country": "Germany",
-                                 "locality": "Berlin",
-                                 "address": "Alexanderplatz 1"}]
-                        can have intermediate stops (> 2 dictionaries within the list)
-                        alternatively param <distance> can be provided
-    :param passengers: Number of passengers in the car (including the person answering the questionnaire),
-                        [1, 2, 3, 4, 5, 6, 7, 8, 9]                             default: 1
-    :param size: size of car
-                        ["small", "medium", "large", "average"]                 default: "average"
-    :param fuel_type: type of fuel the car is using
-                        ["diesel", "gasoline", "cng", "electric", "hybrid", "plug-in_hybrid", "average"]    default: "average"
-    :type distance: float or None
-    :type stops: list[*dict] or None
-    :type passengers: int
-    :type size: str
-    :type fuel_type: str
-    :return: Total emissions of trip in co2 equivalents, total distance of the trip
-    :rtype: tuple[float, float]
+class CalculationError(Exception):
+    """Base exception for all calculation errors"""
+
+
+class InputMissing(CalculationError):
+    """Raised if inputs are not sufficient for calculation"""
+
+
+@enum.unique
+class TransportationMode(str, enum.Enum):
+    """Available transportation modes"""
+
+    CAR = "car"
+    BUS = "bus"
+    TRAIN = "train"
+    PLANE = "plane"
+    FERRY = "ferry"
+
+
+class EmissionRequest(BaseModel):
+    """Generic model for emission calculation request"""
+
+    transportation_mode: TransportationMode
+    distance: Kilometer
+
+
+class Stop(BaseModel):  # Could be renamed to "waypoint"?
+    """Model for a waypoint of a trip"""
+
+    address: str
+    locality: str
+    country: str
+
+
+class StopList(BaseModel):
+    """Model for a list of waypoints"""
+
+    __root__: List[Stop]
+
+    def __iter__(self):
+        return iter(self.__root__)
+
+    def __getitem__(self, item):
+        return self.__root__[item]
+
+
+class DistanceRequest(BaseModel):
+    """Request-model to calculate distances"""
+
+    stops: StopList
+    transportation_mode: TransportationMode
+
+
+def trip_direct(request: DistanceRequest) -> Kilometer:
+    """Function to calculate the distance of a trip as a direct route between locations
+    :param request: validated input data
+    :return: Distance of the trip
+    :rtype: Kilometer
+
     """
-    # NOTE: Tests fail for 'cng'  as `fuel_type` (IndexError)
+    coords = []
+    for loc in DistanceRequest.stops:
+        loc_name, loc_country, loc_coords, _ = geocoding_structured(loc.dict())
+        coords.append(loc_coords)
+    distance = get_route(coords, "driving-car")
```

Comment:
This could be combined with `TransportationMode` enum!


```python

class TransportationMode(str, enum.Enum):
    CAR = "car"
    ...

routing_modes: Dict[TransportationMode, str] = {
    TransportationMode.CAR: "driving-car"
    ...
}

...

distance = get_route(coords, routing_modes[DistanceRequest.transportation_mode])
```

Why?
1) Most likely we can make code even more DRY for different transportation modes
2) If we ever have changes on openrouteservice, we have to replace the string for the mode only once!